src/codec.rs

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Traits to be implemented on structs that can be encoded to or decoded from the wire encoding of
+//! VDAF messages in TLS syntax, as specified in
+//! [RFC 8446, Section 3](https://datatracker.ietf.org/doc/html/rfc8446#section-3).
+
+use std::{
+    io::{Cursor, Read},
+    mem::size_of,
+};
+
+/// Describes how to decode an object from a byte sequence.
+pub trait Decode<D>: Sized {
+    /// Type of errors returned by the [`decode`] method.
+    type Error;
+
+    /// Read and decode an encoded object from `bytes`. `decoding_parameter` provides details of the
+    /// wire encoding such as lengths of different portions of the message.
+    fn decode(decoding_parameter: &D, bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error>;
+}
+
+/// Describes how to encode objects into a byte sequence.
+pub trait Encode {
+    /// Append the encoded form of this object to the end of `bytes`, growing the vector as needed.
+    fn encode(&self, bytes: &mut Vec<u8>);
+}
+
+impl Decode<()> for u8 {
+    type Error = std::io::Error;
```

Comment:
In practice, will we ever have any error besides `std::io::Error`? Even when decoding finite field elements, we'd call `bytes.read_exact()` on a buffer the size of the field element.

Reply:
I think so. You might succeed at reading the bytes, but fail to interpret the bytes you got as the type you're trying to decode into. Obviously it's impossible to fail to interpret a byte into a `u8`, but imagine I was decoding into an enum which is encoded as a `u8` but which only defines 10 variants. The byte `11` is invalid and can't be decoded. Calling that an `std::io::Error` wouldn't be right.

I found that providing an associated type for the error here makes it easier to use this trait. I may want to use the `?` operator in an implementation of `Decode`, and so being able to specify my own error type is helpful because then I can ensure that my error implements all the necessary `From`s.

Reply:
Makes sense.